pandas/core/case_when.py

```diff
@@ -0,0 +1,125 @@
+from __future__ import annotations
+
+from typing import Any
+
+from pandas.core.dtypes.common import is_array_like
+
+import pandas as pd
+import pandas.core.common as com
+
+
+def case_when(obj: pd.DataFrame | pd.Series, *args, default: Any) -> pd.Series:
+    """
+    Returns a Series based on multiple conditions assignment.
+
+    This is useful when you want to assign a column based on multiple conditions.
+    Uses `Series.mask` to perform the assignment.
+
+    Parameters
+    ----------
+    obj : Dataframe or Series on which the conditions will be applied.
+    args : Variable argument of conditions and expected values.
+        Takes the form:
+            `condition0`, `value0`, `condition1`, `value1`, ...
+        `condition` can be a 1-D boolean array/series or a callable
+        that evaluate to a 1-D boolean array/series. See examples below.
+    default : Any
+        The default value to be used if all conditions evaluate False. This value
+        will be used to create the `Series` on which `Series.mask` will be called.
+        If this value is not already an array like (i.e. it is not of type `Series`,
+        `np.array` or `list`) it will be repeated `obj.shape[0]` times in order to
+        create an array like object from it and then apply the `Series.mask`.
+
+    Returns
+    -------
+    Series
+        Series with the corresponding values based on the conditions, their values
+        and the default value.
```

Comment:
I think the result should be guaranteed to have the same index as `obj`. I believe the current implementation doesn't do this - and if that's intentional then it would be useful to see an example.

Reply:
Ah nice catch. Upon looking into this, it turned out we need to maintain the index otherwise some cases will not function as desired (see example below). I fixed it in [cdec601](https://github.com/pandas-dev/pandas/pull/50343/commits/cdec60126dd8e7b3df9b6c7c54ffc47cfdfc9a43). 

Let me know if this does not make "design" sense. 


**Why we need to maintain the index of `obj`?**
Because if we don't, the API cannot support a use case where `obj` is a `Dataframe` and `value0` is a `Series`. To illustrate: 

```
>>> df = pd.DataFrame(dict(a=[1, 2, 3], b=[4, 5, 6]), index=['index 1', 'index 2', 'index 3'])
>>> df
         a  b
index 1  1  4
index 2  2  5
index 3  3  6
>>> pd.case_when(
...     df,
...     lambda x: (x.a == 1) & (x.b == 4),
...     df.b,
...     default=0,
... )
0   NaN
1   NaN
2   NaN
dtype: float64
```

why we get that weird series with `NaN`s instead of the expected series? 

Because the index of the `Series` we apply `mask` on does not match the index of `replacements` and `conditions` parameters passed to `mask` function. 

In simply words, notice how `A` does not match `B` and `C` in the picture below: 
![image](https://user-images.githubusercontent.com/20634657/224178302-4aad3656-c690-4178-956c-4f713eb369b9.png)

Reply:
I realized that with the logic in my comment above, you can still "break" the API as follows: 

```
>>> df = pd.DataFrame(dict(a=[1, 2, 3], b=[4, 5, 6]), index=['index 1', 'index 2', 'index 3'])
>>> df
         a  b
index 1  1  4
index 2  2  5
index 3  3  6
>>> pd.case_when(
...     df,
...     lambda x: (x.a == 1) & (x.b == 4),
...     pd.Series([-1,-1,-1], index=['other index 1', 'other index 1', 'other index 1']),,
...     default=0,
... )
index 1    NaN
index 2    0.0
index 3    0.0
dtype: float64
```

With that being said, instead of adding `assert`s to check for matching indices before calling `Series.mask`, I opted for the simple solution of resetting the index for now. See [df1845d](https://github.com/pandas-dev/pandas/pull/50343/commits/df1845d6931293070ea1acbb43697351897f2381). 

Like this, the following use case will not be supported out of the box with this API and the user will need to manually handle the indices (which I think is fine at least for the first iteration of this API): 

```
>>> df = pd.DataFrame(dict(a=[1, 2, 3], b=[4, 5, 6]), index=['index 1', 'index 2', 'index 3'])
>>> df
         a  b
index 1  1  4
index 2  2  5
index 3  3  6
>>> df.assign(
...     new_col = case_when(
...         df,
...         lambda x: (x.a == 1) & (x.b == 4),
...         df.b,
...         default=0,
...     )
... )
         a  b  new_col
index 1  1  4      NaN
index 2  2  5      NaN
index 3  3  6      NaN
```